pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -242,3 +319,124 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
 	}
 	return nil
 }
+
+func (c *kvEventToRowConsumer) Close() error {
+	return nil
+}
+
+type parallelEventConsumer struct {
+	g            ctxgroup.Group
+	spawnAfter   *timeutil.Timer
+	eventCh      chan kvevent.Event
+	termCh       chan struct{}
+	doneCh       chan struct{}
+	maxWorkers   func() int64
+	makeConsumer func() (eventConsumer, error)
+
+	mu struct {
+		syncutil.Mutex
+		numWorkers int64
+		termErr    error
+	}
+}
+
+var _ eventConsumer = (*parallelEventConsumer)(nil)
+
+const spawnWorkerDelay = 50 * time.Millisecond
+
+func (c *parallelEventConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Event) error {
+	c.spawnAfter.Reset(spawnWorkerDelay)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-c.termCh:
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			return c.mu.termErr
+		case c.eventCh <- ev:
+			return nil
+		case <-c.spawnAfter.C:
+			c.spawnAfter.Read = true
+			if err := c.trySpawn(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (c *parallelEventConsumer) trySpawn() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.mu.numWorkers > c.maxWorkers() {
+		return nil
+	}
+
+	c.mu.numWorkers++
+	c.g.GoCtx(c.workerLoop)
+	return nil
+}
+
+// jitter adds a small jitter in the given duration.
+func idleAfter() time.Duration {
+	const idleAfter = 2 * time.Minute
+	const jitter = 1.0 / 6.0
+	jitterFraction := 1 + (2*rand.Float64()-1)*jitter // 1 + [-1/6, +1/6)
+	return time.Duration(float64(idleAfter) * jitterFraction)
+}
+
+func (c *parallelEventConsumer) workerLoop(ctx context.Context) error {
+	consumer, err := c.makeConsumer()
+	if err != nil {
+		return c.setWorkerError(err)
+	}
+
+	idleTimer := timeutil.NewTimer()
+	defer idleTimer.Stop()
+
+	for {
+		idleTimer.Reset(idleAfter())
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-c.doneCh:
+			return nil
+		case e := <-c.eventCh:
+			if err := consumer.ConsumeEvent(ctx, e); err != nil {
+				return c.setWorkerError(err)
+			}
+		case <-idleTimer.C:
+			idleTimer.Read = true
+
+			c.mu.Lock()
+			shutDown := c.mu.numWorkers > 1
```

Comment:
Is it that bad to leave extra workers idle?  My assumption would've been that having an extra goroutine idling around is cheap.

Reply:
sure; but ... why...

Reply:
My question is "why do extra work / have extra constraints for dynamic number of workers if it isn't providing value", where the extra work is the code around ramping up / down workers, the extra constraints is being slightly slower to ramp up capacity, while the value is something I'm not clear on since the only cost saving I can think of is having an extra idle routine.

Reply:
> My question is "why do extra work / have extra constraints for dynamic number of workers if it isn't providing value", where the extra work is the code around ramping up / down workers, the extra constraints is being slightly slower to ramp up capacity, while the value is something I'm not clear on since the only cost saving I can think of is having an extra idle routine.

I suppose this can be discussed once review (proper) is started...
Note however: it's not just about go routines; each one creates encoder, fetcher, "evaluator" -- these are
large datastructures, sometimes having their own caches, etc.

Timers are also cheap, and so is shutdown if worker didn't see any work in 2 minutes.